### ISSUE TITLE
Classification of difficulty is not good enough

### ISSUE THREAD

**Issue by user1823** [open-spaced-repetition/fsrs4anki#262]
In my collection, a large number of cards have been assigned the difficulty of 10 (the maximum possible number). The actual distribution of difficulty for my collection is shown in the screenshot below.

<img width=256px src="https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/b148c62c-f379-46a7-a16a-e80485a284ae">

The complete analysis can be seen in [prediction.tsv.zip](https://github.com/open-spaced-repetition/fsrs4anki/files/11508098/prediction.tsv.zip)

However, I am sure that these cards are not equally difficult. Some of these are much more difficult than the others. You can also see in the prediction.tsv that the review history of cards with difficulty of 10 is quite varied.

So, this means that FSRS doesn't accurately classify the cards based on their difficulty.

This is a problem because when heterogeneous materials are assigned the same difficulty, it is not possible to obtain accurate values of the parameters that are used to calculate the next interval for these cards.

In other words, what is the point of defining difficulty if the cards are going to be clustered at only a few difficulty levels?

I am creating a separate issue for this instead of adding this to https://github.com/open-spaced-repetition/fsrs4anki/issues/248 (the "add retrievability in difficulty" issue) because the current issue is not necessarily associated with cards having low retrievability at the time of review.

Other information:
- The above analysis was generated using the default optimizer, not any of the experimental versions.
- w = [1.0452, 1.6989, 4.9877, -1.26, -1.1786, 0.0, 1.7037, -0.0941, 1.0638, 1.7052, -0.4985, 0.8364, 0.5216]

**Comment by L-M-Sherlock** [comment]
It is hard to say whether the difficulty classification is good or not. Do you have the distribution of difficulty in SuperMemo? I only find the distribution of the A-factor in woz's wiki:

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/1adb092d-3182-4655-b832-ea1b639e6880)

Source: https://help.supermemo.org/wiki/Analysis#Distributions

By the way, I also use SuperMemo. It is my distribution of difficutly:

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/14bb73f0-51ba-4979-bbe2-bfc38350fed4)

**Comment by L-M-Sherlock** [comment]
I need some cases for further research. Could you give some examples whose difficulty is same but you think they should be significantly different?

**Comment by user1823** [comment]
> I need some cases for further research. Could you give some examples whose difficulty is same but you think they should be significantly different?

Consider the selected rows in the following screenshots:

<img width=500px src="https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/51337bfb-5a14-4c29-8bd3-8265b9cddc36">
<img width=1000px src="https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/34a36fe8-8536-4219-ac02-cc1d7a0dacdb">
<img width=1000px src="https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/64fa5c03-54e5-4467-814a-f1a94c64821d">
<img width=1000px src="https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/0c1417f3-070d-464f-8c26-2c4136685726">
<img width=1000px src="https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/67934c28-e70a-4a88-8e6f-7c5c1485362f">
<img width=1000px src="https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/c7f0daea-e262-4dfb-9911-227d58301b60">
<img width=1000px src="https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/89830ce3-5599-4291-b035-8a6f1dcd02cc">
<img width=1000px src="https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/f60bfa11-24dc-4c63-855a-b6896d30d8e1">

These have very different review histories but still have the same difficulty.

You can find more examples in [prediction.tsv.zip](https://github.com/open-spaced-repetition/fsrs4anki/files/11508098/prediction.tsv.zip) (rename it to remove .zip at the end).

**Comment by user1823** [comment]
I think that this issue is caused by the difficulty function in FSRS being oversimplified.

For example, I think that we should revisit how the grades affect the difficulty because (G-3) seems oversimplified. Why should Good have no effect on the difficulty? As a band-aid solution for the problem of difficulty hell that results from this formula, you implemented the mean_reversion function. But why should the difficulty always move towards the initial value? This seems a bit ad-hoc.

The following suggests that SuperMemo uses the BW metric to determine the difficulty. But, then how does it take into account the grade? We need to think about that.
> Algorithm SM-17 would look for absolute item difficulty that would provide the best fit to the memory model (minimum deviation of grades from the expected recall). Algorithm SM-18 takes each repetition separately, and estimates expectation-based difficulty for that repetition only. It maps linearly from bet-win metric (BW) to difficulty in the range from 0.0 to 1.0 (see the picture). As the default forgetting index is 10%, all items with BW above 0.1 are considered maximally easy (1.0-0.9=0.1), and all items with BW below -0.9 are considered maximally hard (0.0-0.9=-0.9).

Source: https://supermemo.guru/wiki/Item_difficulty_in_Algorithm_SM-18

**Comment by user1823** [comment]
By the way, @L-M-Sherlock, what is your opinion on having "improving the functions for calculating difficulty in spaced repetition algorithms" as the topic of your next research?

I think that changing how difficulty is calculated is the most promising area for further improvement of the algorithm.

**Comment by L-M-Sherlock** [comment]
https://github.com/open-spaced-repetition/fsrs4anki/blob/Expt/new-baseline/experiment/mini-batch_power-difficulty.ipynb

I think this classification of difficulty is good:

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/be0f9522-3856-4240-b414-13a4e18d625b)

Note: the backlog in difficulty = 1 is caused by my rating habits. I press `easy` a lot.

**Comment by L-M-Sherlock** [comment]
After https://github.com/open-spaced-repetition/fsrs4anki/pull/283, the difficulty will be distributed more uniformly.

![image](https://user-images.githubusercontent.com/32575846/243747318-c7bea8a4-594b-49eb-9116-0bac392abb04.png)

Could you test it? @user1823

**Comment by Expertium** [comment]
I will test it later today, but I'm willing to bet it won't improve performance.

**Comment by L-M-Sherlock** [comment]
> I will test it later today, but I'm willing to bet it won't improve performance.

It would not affect you. Because your w[5] is larger than 0.05.

**Comment by user1823** [comment]
> After #283, the difficulty will be distributed more uniformly.
> 
> ![image](https://user-images.githubusercontent.com/32575846/243747318-c7bea8a4-594b-49eb-9116-0bac392abb04.png)
> 
> Could you test it? @user1823

After this patch, the distribution of difficulty is undoubtedly more uniform. But, the loss and RMSE have increased slightly.

Before:
Optimized w[5] = 0
Loss = 0.2220
RMSE = 0.0175

Now:
Optimized w[5] = 0.05
Loss = 0.2247
RMSE = 0.0183

**Comment by user1823** [comment]
I am closing this because I am creating a new issue combining all problems related to the calculation of difficulty.
